overlord/exportstate/special.go

```diff
@@ -0,0 +1,187 @@
+		SourceIsHost:              true,
+		Sets:                      map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForSnapdSnap returns the manifest of the snapd snap.
+func manifestForSnapdSnap(info *snap.Info) *Manifest {
+	tools := snapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision,
+		Sets:             map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForCoreSnap returns the manifest of the core snap.
+func manifestForCoreSnap(info *snap.Info) *Manifest {
+	tools := snapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName:          info.InstanceName(),
+		SnapRevision:              info.Revision,
+		ExportedForSnapdAsVersion: fmt.Sprintf("core_%s", info.Revision), // Exception from the rule
+
+		// Separate to avoid gofmt annoyance across versions.
+		Sets: map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForRegularSnap returns the manifest for a snap other than core or snapd.
+func manifestForRegularSnap(info *snap.Info) *Manifest {
+	return &Manifest{
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision,
+		// TODO: eventually get this from the snap.yaml
+	}
+}
+
+func exportedNameVersionForRegularSnap(info *snap.Info) (exportedName string, exportedVersion string) {
+	if typ := info.Type(); typ == snap.TypeSnapd || typ == snap.TypeOS {
+		panic(fmt.Sprintf("internal error, cannot use exportedNameVersionForRegularSnap with core or snapd, got %q of type %q", info.SnapName(), info.Type()))
+	}
+	return info.InstanceName(), info.SnapRevision().String()
+}
+
+func effectiveExportedVersionForSnapdOrCore(st *state.State) (exportedVersion string, err error) {
+	snapdInfo, coreInfo, err := currentSnapdAndCoreInfo(st)
+	if err != nil {
+		return "", err
+	}
+	var activeSnapdExportedVersion string
+	var activeCoreExportedVersion string
+	if snapdInfo != nil && snapdInfo.Broken == "" {
+		activeSnapdExportedVersion = snapdInfo.Revision.String()
+	}
+	if coreInfo != nil && coreInfo.Broken == "" {
+		activeCoreExportedVersion = fmt.Sprintf("core_%s", coreInfo.Revision)
+	}
+	exportedVersion = selectExportedVersionForSnapdTools(activeSnapdExportedVersion, activeCoreExportedVersion)
+	return exportedVersion, nil
+}
+
+// selectExportedVersionForSnapdTools returns the version to use for snapd tools export set.
+//
+// The snapd tools export set is special as there are providers from snaps other
+// than snapd that need consideration. The result is, in order of preference:
+//
+// 0) "host" if on classic with disabled re-execution.
+// 1) snapd version, if available
+// 2) core version, if available
+// 3) "host" version, if on classic
+//
+// If no provider is available then empty version is returned.
+func selectExportedVersionForSnapdTools(activeSnapdExportedVersion, activeCoreExportedVersion string) string {
+	if release.OnClassic && os.Getenv("SNAP_REEXEC") == "0" {
```

Comment:
shouldn't this really use snapdtool.IsReexecd instead of just looking at the var like this?  the var actually will be often just unset

Reply:
I honestly don't like that, it's a special case that's distributed. This mode is really nicer, even if it means the symlinks can point to non-snap places. As another example of non-snap places that symlinks _could_ point to are the nvidia libraries we are reusing from the host.

Reply:
but the reality is that SNAP_REEXEC is a env var that can be set differently for single invocations

Reply:
In the interest of stability I would keep this as-is, review the C side of this and the spread tests and _then_ change how SNAP_REEXEC is used (client side).

Reply:
anyway if we don't make that change I still think this needs to use for the time being snapdtool.IsReexecd no?